File: frontend/app/types/inquiry.ts
/**
 * Inquiry API Types - matches backend reee/inquiry/types.py
 */

// Enums
export type InquiryStatus = 'open' | 'resolved' | 'stale' | 'closed'
export type RigorLevel = 'A' | 'B' | 'C'
export type ContributionType = 'evidence' | 'refutation' | 'attribution' | 'scope_correction' | 'disambiguation'
export type TaskType = 'need_primary_source' | 'unresolved_conflict' | 'single_source_only' | 'high_entropy' | 'stale'
export type ObservationKind = 'point' | 'lower_bound' | 'upper_bound' | 'interval' | 'approximate' | 'none'
export type SchemaType = 'monotone_count' | 'categorical' | 'boolean' | 'report_truth' | 'quote_authenticity' | 'custom' | 'forecast'

// Core Types
export interface InquirySchema {
  schema_type: SchemaType
  categories?: string[]
  count_scale?: 'small' | 'medium' | 'large'
  count_max?: number
  count_monotone?: boolean
  hypotheses?: string[]
  rigor: RigorLevel
}

export interface InquirySummary {
  id: string
  title: string
  status: InquiryStatus
  rigor: RigorLevel
  schema_type: SchemaType
  posterior_map: any
  posterior_probability: number
  entropy_bits: number
  stake: number  // "bounty" in UI
  contributions: number
  open_tasks: number
  resolvable: boolean
  resolved_ago?: string
  scope_entities?: string[]
  cover_image?: string
  credible_interval?: [number, number]
}

export interface InquiryDetail extends InquirySummary {
  description?: string
  scope_keywords?: string[]
  scope_time_start?: string
  scope_time_end?: string
  normalized_entropy: number
  created_at: string
  updated_at: string
  created_by?: string
  blocking_tasks: string[]
}

// Observation (from TypedBeliefState)
export interface Observation {
  kind: ObservationKind | string  // Allow string for simulated data
  value_distribution: Record<string | number, number>
  bound_value?: number
  interval_low?: number
  interval_high?: number
  source: string
  claim_id?: string
  timestamp?: number
  extraction_confidence?: number  // Optional for simulated
  signals?: Record<string, any>
}

// Contribution (user submission)
export interface Contribution {
  id: string
  inquiry_id?: string  // Optional for simulated
  user_id?: string
  user_name?: string
  type: ContributionType | string  // Allow string for simulated
  text: string
  source_url?: string
  source_name?: string
  source?: string  // Alternative to source_name
  timestamp?: string
  extracted_value?: any
  observation_kind?: ObservationKind | string
  created_at: string
  processed?: boolean  // Optional for simulated
  claim_ids?: string[]
  posterior_impact?: number  // Optional for simulated
  impact?: number  // Alternative field name
}

// Task (from meta-claims)
export interface InquiryTask {
  id: string
  inquiry_id: string
  type: TaskType
  description: string
  bounty: number
  created_at: string
  claimed_by?: string
  completed: boolean
  completed_at?: string
  meta_claim_id?: string
}

// Surface (L2 cluster)
export interface Surface {
  id: string
  name: string
  claim_count: number
  sources: string[]
  entities?: string[]
  in_scope: boolean
  entropy?: number
  canonical_value?: any
  relations: SurfaceRelation[]
}

export interface SurfaceRelation {
  type: 'CONFIRMS' | 'SUPERSEDES' | 'CONFLICTS' | 'REFINES'
  target: string
  confidence?: number
}

// Belief State (posterior summary)
export interface BeliefState {
  map: any
  map_probability: number
  entropy_bits: number
  normalized_entropy: number
  observation_count: number
  log_scores?: number[]
  total_log_score: number
}

// Resolution status
export interface Resolution {
  status: InquiryStatus
  resolvable: boolean
  stable_since?: string
  blocking_tasks: string[]
  hours_stable?: number
}

// Full trace response
export interface InquiryTrace {
  inquiry: { id: string; title: string } | InquirySummary
  belief_state: BeliefState
  observations: Observation[]
  contributions: Contribution[]
  surfaces?: Surface[]
  tasks: InquiryTask[]
  resolution: Resolution
  posterior_top_10?: Array<{ value: any; probability: number }>
  claims?: Array<{
    id: string
    icon: string
    source: string
    text: string
    extracted_value: any
    observation_kind: string
  }>
}

// Replay state (for animation)
export interface ReplayState {
  currentIndex: number
  isPlaying: boolean
  speed: number // 1x, 2x, etc.
  snapshots: BeliefSnapshot[]
}

export interface BeliefSnapshot {
  index: number
  contribution: Contribution
  belief_state: BeliefState
  posterior_top_10: Array<{ value: any; probability: number }>
  entropy_delta: number
  probability_delta: number
}

// API Input types
export interface ContributionInput {
  type: ContributionType
  text: string
  source_url?: string
  source_name?: string
  extracted_value?: any
  observation_kind?: ObservationKind
  // For attribution
  attributed_to?: string
  original_source?: string
  // For scope correction
  correct_scope?: string
  out_of_scope_reason?: string
}

export interface StakeInput {
  amount: number
}

export interface CreateInquiryInput {
  title: string
  description?: string
  schema: InquirySchema
  scope_entities?: string[]
  scope_keywords?: string[]
  scope_time_start?: string
  scope_time_end?: string
  initial_stake?: number
}

// Alias for convenience
export type Task = InquiryTask

File: backend/models/api/inquiry.py
"""
Inquiry API Models (Pydantic schemas)

Request/Response models for the Inquiry MVP API.
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Any, Dict
from datetime import datetime
from enum import Enum


# =============================================================================
# ENUMS FOR VALIDATION
# =============================================================================

class InquiryStatusEnum(str, Enum):
    OPEN = "open"
    RESOLVED = "resolved"
    STALE = "stale"
    CLOSED = "closed"


class RigorLevelEnum(str, Enum):
    A = "A"
    B = "B"
    C = "C"


class SchemaTypeEnum(str, Enum):
    MONOTONE_COUNT = "monotone_count"
    CATEGORICAL = "categorical"
    BOOLEAN = "boolean"
    REPORT_TRUTH = "report_truth"
    QUOTE_AUTHENTICITY = "quote_authenticity"
    FORECAST = "forecast"
    CUSTOM = "custom"


class ContributionTypeEnum(str, Enum):
    EVIDENCE = "evidence"
    REFUTATION = "refutation"
    ATTRIBUTION = "attribution"
    SCOPE_CORRECTION = "scope_correction"
    DISAMBIGUATION = "disambiguation"


class TaskTypeEnum(str, Enum):
    NEED_PRIMARY_SOURCE = "need_primary_source"
    UNRESOLVED_CONFLICT = "unresolved_conflict"
    SINGLE_SOURCE_ONLY = "single_source_only"
    HIGH_ENTROPY = "high_entropy"
    STALE = "stale"


class ObservationKindEnum(str, Enum):
    POINT = "point"
    LOWER_BOUND = "lower_bound"
    UPPER_BOUND = "upper_bound"
    INTERVAL = "interval"
    APPROXIMATE = "approximate"
    NONE = "none"


# =============================================================================
# REQUEST MODELS
# =============================================================================

class InquirySchemaInput(BaseModel):
    """Schema definition for new inquiry."""
    schema_type: SchemaTypeEnum = SchemaTypeEnum.BOOLEAN
    categories: List[str] = []
    count_scale: str = "medium"
    count_max: int = 500
    count_monotone: bool = True
    hypotheses: List[str] = []
    rigor: RigorLevelEnum = RigorLevelEnum.B


class CreateInquiryInput(BaseModel):
    """Input for creating an inquiry."""
    title: str = Field(..., min_length=10, max_length=500, description="The question to investigate")
    description: str = Field("", max_length=2000, description="Additional context")
    schema: InquirySchemaInput = InquirySchemaInput()
    scope_entities: List[str] = Field([], description="Required entities for scope")
    scope_keywords: List[str] = Field([], description="Keywords to match")
    scope_time_start: Optional[datetime] = None
    scope_time_end: Optional[datetime] = None
    initial_stake: float = Field(0.0, ge=0, description="Initial bounty stake")


class ContributionInput(BaseModel):
    """Input for adding a contribution."""
    type: ContributionTypeEnum = ContributionTypeEnum.EVIDENCE
    text: str = Field(..., min_length=10, max_length=2000, description="Quote or description")
    source_url: Optional[str] = Field(None, max_length=2000, description="Source URL")
    source_name: Optional[str] = Field(None, max_length=255, description="Source name")
    extracted_value: Optional[Any] = Field(None, description="Extracted value for typed inquiries")
    observation_kind: ObservationKindEnum = ObservationKindEnum.POINT


class StakeInput(BaseModel):
    """Input for adding stake."""
    amount: float = Field(..., gt=0, description="Stake amount in credits")


class ClaimTaskInput(BaseModel):
    """Input for claiming a task."""
    pass  # Just need the task_id from path


class CompleteTaskInput(BaseModel):
    """Input for completing a task."""
    contribution_id: Optional[str] = Field(None, description="ID of contribution that completes task")


# =============================================================================
# RESPONSE MODELS
# =============================================================================

class InquirySummary(BaseModel):
    """Summary response for inquiry (list view)."""
    id: str
    title: str
    status: str
    rigor: str
    schema_type: str
    posterior_map: Optional[Any] = None
    posterior_probability: float
    entropy_bits: float
    stake: float
    contributions: int
    open_tasks: int
    resolvable: bool
    scope_entities: List[str] = []
    cover_image: Optional[str] = None
    resolved_ago: Optional[str] = None

    class Config:
        from_attributes = True


class BeliefStateResponse(BaseModel):
    """Belief state summary."""
    map: Optional[Any] = None
    map_probability: float
    entropy_bits: float
    normalized_entropy: float
    observation_count: int
    total_log_score: float


class InquiryDetail(BaseModel):
    """Detailed response for inquiry."""
    id: str
    title: str
    description: str
    status: str
    rigor: str
    schema_type: str
    scope_entities: List[str]
    scope_keywords: List[str]
    posterior: BeliefStateResponse
    stake: float
    contributions: int
    open_tasks: int
    created_at: str
    updated_at: str
    resolvable: bool
    credible_interval: Optional[List[Any]] = None

    class Config:
        from_attributes = True


class ContributionResponse(BaseModel):
    """Response for a contribution."""
    id: str
    inquiry_id: str
    user_id: Optional[str] = None
    user_name: str = "Anonymous"
    type: str
    text: str
    source_url: Optional[str] = None
    source_name: Optional[str] = None
    extracted_value: Optional[Any] = None
    observation_kind: Optional[str] = None
    processed: bool
    impact: float
    posterior_impact: float
    reward_earned: float
    created_at: str

    class Config:
        from_attributes = True


class TaskResponse(BaseModel):
    """Response for a task."""
    id: str
    inquiry_id: str
    type: str
    description: str
    bounty: float
    claimed_by: Optional[str] = None
    claimed_by_name: Optional[str] = None
    completed: bool
    created_at: str

    class Config:
        from_attributes = True


class StakeResponse(BaseModel):
    """Response for a stake."""
    user_id: str
    user_name: str
    amount: float
    created_at: str


class TransactionResponse(BaseModel):
    """Response for a credit transaction."""
    id: int
    amount: float
    balance_after: float
    transaction_type: str
    reference_type: Optional[str] = None
    reference_id: Optional[str] = None
    description: Optional[str] = None
    created_at: str


class ResolutionResponse(BaseModel):
    """Resolution status response."""
    status: str
    resolvable: bool
    stable_since: Optional[str] = None
    hours_stable: Optional[float] = None
    blocking_tasks: List[str] = []


class TraceResponse(BaseModel):
    """Full epistemic trace response."""
    inquiry: InquirySummary
    belief_state: BeliefStateResponse
    contributions: List[ContributionResponse]
    tasks: List[TaskResponse]
    stakes: List[StakeResponse]
    resolution: ResolutionResponse
    posterior_top_10: Optional[List[Dict[str, Any]]] = None
    surfaces: Optional[List[Dict[str, Any]]] = None


class ContributionResultResponse(BaseModel):
    """Response after adding a contribution."""
    contribution: ContributionResponse
    updated_posterior: BeliefStateResponse


class StakeResultResponse(BaseModel):
    """Response after adding a stake."""
    total_stake: float
    user_balance: float


class UserCreditsResponse(BaseModel):
    """Response for user credits."""
    user_id: str
    credits_balance: float
    recent_transactions: List[TransactionResponse] = []
